File: examples/java/docs/PrivateOrderServiceGetOrderEvents.java
/*
 * Copyright 2019 Tulip Solutions B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package docs;

import io.grpc.ManagedChannel;
import io.grpc.stub.StreamObserver;
import nl.tulipsolutions.api.common.SearchDirection;
import nl.tulipsolutions.api.common.Market;
import nl.tulipsolutions.api.priv.CancelOrderEvent;
import nl.tulipsolutions.api.priv.CreateLimitOrderEvent;
import nl.tulipsolutions.api.priv.FillOrderEvent;
import nl.tulipsolutions.api.priv.GetEventsForOrderRequest;
import nl.tulipsolutions.api.priv.GetOrderEventsRequest;
import nl.tulipsolutions.api.priv.GetOrderEventsResponse;
import nl.tulipsolutions.api.priv.OrderEvent;
import nl.tulipsolutions.api.priv.PrivateOrderServiceGrpc;
import nl.tulipsolutions.api.priv.PrivateOrderServiceGrpc.PrivateOrderServiceStub;

import java.util.concurrent.TimeUnit;

public class PrivateOrderServiceGetOrderEvents {
    public static void run(ManagedChannel channel) {
        // CODEINCLUDE-BEGIN-MARKER: ref-code-example-request
        PrivateOrderServiceStub stub = PrivateOrderServiceGrpc.newStub(channel);

        // Create a request for 10 most recent order events on the BTC_EUR and BTC_USD markets
        GetOrderEventsRequest request = GetOrderEventsRequest.newBuilder()
            .addMarkets(Market.BTC_EUR)
            .addMarkets(Market.BTC_USD)
            .setSearchDirection(SearchDirection.BACKWARD)
            .setLimit(10)
            .build();

        // Make the request asynchronously with a one second deadline
        stub.withDeadlineAfter(1, TimeUnit.SECONDS)
            .getOrderEvents(request, new StreamObserver<GetOrderEventsResponse>() {
                public void onNext(GetOrderEventsResponse response) {
                    System.out.println(response);
                    // CODEINCLUDE-END-MARKER: ref-code-example-request
                    parseAndPrint(response);
                    // CODEINCLUDE-BEGIN-MARKER: ref-code-example-request
                }

                public void onError(Throwable t) {
                    System.err.println("PrivateOrderService.GetOrderEvents error: " + t.getMessage());
                }

                public void onCompleted() {
                    System.out.println("PrivateOrderService.GetOrderEvents completed");
                }
            });
        // CODEINCLUDE-END-MARKER: ref-code-example-request
    }

    public static void parseAndPrint(GetOrderEventsResponse response) {
        // CODEINCLUDE-BEGIN-MARKER: ref-code-example-response
        StringBuilder formattedResponse = new StringBuilder(String.format("%s\n", response.getClass().getSimpleName()));
        for (OrderEvent event : response.getEventsList()) {
            String formattedEvent;
            switch (event.getEventCase()) {
                case CREATE_LIMIT_ORDER_EVENT:
                    CreateLimitOrderEvent limitOrder = event.getCreateLimitOrderEvent();
                    formattedEvent = String.format(
                        "\t%s: Event %d order %d on market %s %s %f@%f\n",
                        limitOrder.getClass().getSimpleName(),
                        event.getEventId(),
                        event.getOrderId(),
                        event.getMarket(),
                        limitOrder.getSide(),
                        limitOrder.getBaseAmount(),
                        limitOrder.getPrice()
                    );
                    break;
                case FILL_ORDER_EVENT:
                    FillOrderEvent fillOrderEvent = event.getFillOrderEvent();
                    formattedEvent = String.format(
                        "\t%s: Event %d order %d on market %s %s %f@%f\n",
                        fillOrderEvent.getClass().getSimpleName(),
                        event.getEventId(),
                        event.getOrderId(),
                        event.getMarket(),
                        fillOrderEvent.getSide(),
                        fillOrderEvent.getBaseAmount(),
                        fillOrderEvent.getPrice()
                    );
                    break;
                case CANCEL_ORDER_EVENT:
                    CancelOrderEvent cancelOrderEvent = event.getCancelOrderEvent();
                    formattedEvent = String.format(
                        "\t%s: Event %d order %d on market %s\n",
                        cancelOrderEvent.getClass().getSimpleName(),
                        event.getEventId(),
                        event.getOrderId(),
                        event.getMarket()
                    );
                    break;
                default:
                    formattedEvent = "";
                    break;
            }

            formattedResponse.append(formattedEvent);
        }
        System.out.println(formattedResponse);
        // CODEINCLUDE-END-MARKER: ref-code-example-response
    }
}

File: examples/node/docs/privateOrderServiceGetOrderEvents.js
/*
 * Copyright 2019 Tulip Solutions B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
const util = require('util');

const order_pb = require('@tulipsolutions/tecl/priv/order_pb');
const orders_pb = require('@tulipsolutions/tecl/common/orders_pb');
const order_grpc_pb = require('@tulipsolutions/tecl/priv/order_grpc_pb');

function privateOrderServiceGetOrderEvents(host, credentials, options) {
  // CODEINCLUDE-BEGIN-MARKER: ref-code-example-request
  const client = new order_grpc_pb.PrivateOrderServiceClient(host, credentials);

  // Create a request for 10 most recent order events on the BTC_EUR and BTC_USD markets
  const request = new order_pb.GetOrderEventsRequest();
  request.addMarkets(orders_pb.Market.BTC_EUR);
  request.addMarkets(orders_pb.Market.BTC_USD);
  request.setSearchDirection(orders_pb.SearchDirection.BACKWARD);
  request.setLimit(10);

  // Add a 1s deadline, and make the request asynchronously
  const deadline = new Date().setSeconds(new Date().getSeconds() + 1);
  const callOptions = Object.assign({ deadline: deadline }, options);
  client.getOrderEvents(request, callOptions, (err, response) => {
    if (err) {
      console.error('PrivateOrderService.GetEventsForOrder error: ' + err.message);
    }
    if (response) {
      console.log(response.toObject());
      // CODEINCLUDE-END-MARKER: ref-code-example-request
      parseAndPrint(response);
      // CODEINCLUDE-BEGIN-MARKER: ref-code-example-request
    }
  });
  // CODEINCLUDE-END-MARKER: ref-code-example-request
}

function parseAndPrint(response) {
  // CODEINCLUDE-BEGIN-MARKER: ref-code-example-response
  let result = util.format('%s\n', 'GetOrderEventsResponse');
  response.getEventsList().forEach(event => {
    switch (event.getEventCase()) {
      case order_pb.OrderEvent.EventCase.CREATE_ORDER_EVENT: {
        const createOrderEvent = event.getCreateOrderEvent();
        let deadline = '(no deadline)';
        if (createOrderEvent.getDeadlineNs() != 0) {
          deadline = util.format('deadline @ %d', createOrderEvent.getDeadlineNs());
        }
        switch (createOrderEvent.getOrderTypeCase()) {
          case order_pb.CreateOrderEvent.OrderTypeCase.CREATE_LIMIT_ORDER: {
            const limitOrderEvent = createOrderEvent.getCreateLimitOrder();
            result += util.format(
              '\t%s: Event %d order %d on market %s %s %f@%f %s\n',
              'CreateLimitOrderEvent',
              event.getEventId(),
              event.getOrderId(),
              Object.keys(orders_pb.Market).find(key => orders_pb.Market[key] === event.getMarket()),
              Object.keys(orders_pb.Side).find(key => orders_pb.Side[key] === limitOrderEvent.getSide()),
              limitOrderEvent.getBaseAmount(),
              limitOrderEvent.getPrice(),
              deadline
            );
            break;
          }
        }
        break;
      }
      case order_pb.OrderEvent.EventCase.FILL_ORDER_EVENT: {
        const fillOrderEvent = event.getFillOrderEvent();
        result += util.format(
          '\t%s: Event %d order %d on market %s %s %f@%f\n',
          'FillOrderEvent',
          event.getEventId(),
          event.getOrderId(),
          Object.keys(orders_pb.Market).find(key => orders_pb.Market[key] === event.getMarket()),
          Object.keys(orders_pb.Side).find(key => orders_pb.Side[key] === fillOrderEvent.getSide()),
          fillOrderEvent.getBaseAmount(),
          fillOrderEvent.getPrice()
        );
        break;
      }
      case order_pb.OrderEvent.EventCase.CANCEL_ORDER_EVENT:
        result += util.format(
          '\t%s: Event %d order %d on market %s\n',
          'CancelOrderEvent',
          event.getEventId(),
          event.getOrderId(),
          Object.keys(orders_pb.Market).find(key => orders_pb.Market[key] === event.getMarket())
        );
        break;
    }
  });
  console.log(result);
  // CODEINCLUDE-END-MARKER: ref-code-example-response
}

module.exports = privateOrderServiceGetOrderEvents;
